src/snabbdom.ts

```diff
@@ -1,55 +1,90 @@
-/* global module, document, Node */
-import {Module} from './modules/module';
-import {Hooks} from './hooks';
-import vnode, {VNode, VNodeData, Key} from './vnode';
-import * as is from './is';
-import htmlDomApi, {DOMAPI} from './htmldomapi';
-
-function isUndef(s: any): boolean { return s === undefined; }
-function isDef(s: any): boolean { return s !== undefined; }
+/* global Node */
+import {Hooks, InsertHook, VNode, VNodes, Key, VText, VComment,
+        Module, VElement, VFragment, DOMAPI, PreHook, CreateHook, UpdateHook, DestroyHook, RemoveHook, PostHook} from './types';
 
-type VNodeQueue = Array<VNode>;
+const { isArray } = Array;
 
-const emptyNode = vnode('', {}, [], undefined, undefined);
+const ELEMENT_NODE = 1, TEXT_NODE = 3, COMMENT_NODE = 8, DOCUMENT_FRAGMENT_NODE = 11;
 
-function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
-  return vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
-}
+function isUndef(s: any): s is undefined { return s === undefined; }
+function isDef(s: any): boolean { return s !== undefined; }
 
-function isVnode(vnode: any): vnode is VNode {
-  return vnode.sel !== undefined;
-}
+type VNodeQueue = VNode[];
+
+const emptyNode: VNode = {
+  nt: 0,
+  sel: '',
+  data: {},
+  children: undefined,
+  text: undefined,
+  elm: undefined,
+  key: undefined,
+};
 
 type KeyToIndexMap = {[key: string]: number};
 
 type ArraysOf<T> = {
   [K in keyof T]: (T[K])[];
 }
 
-type ModuleHooks = ArraysOf<Module>;
+export interface PatchFunction {
```

Comment:
the patch function returned by `init` is now more restrictive, it only accepts vnodes since snabbdom doesn't really know about the DOM anymore. By default, you can rely on `toVNode()` as the first argument. This removed some weird behavior that result in not spidering the whole structure of the provided node, which eventually let old DOM elements hanging there after the patch.

additionally, a `children` member is introduced to only patch the children collection without touching the parent node at all. This is mostly to support shadow root, and fragments, but I'm ok with removing this if you feel this is too much.